src/Sugar.ml

```diff
@@ -180,45 +180,43 @@ let infix cmts prec xexp =
   in
   infix_ None ~relocate:false (Nolabel, xexp)
 
-let rec list_pat cmts pat =
-  let ctx = Pat pat in
-  let {ppat_desc; ppat_loc= src} = pat in
-  match ppat_desc with
-  | Ppat_construct ({txt= Lident "[]"; loc}, None) ->
-      Cmts.relocate cmts ~src ~before:loc ~after:loc ;
-      Some ([], loc)
-  | Ppat_construct
-      ( {txt= Lident "::"; loc}
-      , Some {ppat_desc= Ppat_tuple [hd; tl]; ppat_loc; ppat_attributes= []}
-      ) -> (
-    match list_pat cmts tl with
-    | Some (xtl, nil_loc) when List.is_empty tl.ppat_attributes ->
-        Some (([src; loc; ppat_loc], sub_pat ~ctx hd) :: xtl, nil_loc)
-    | _ -> None )
-  | _ -> None
+let list_pat cmts pat =
+  let rec list_pat_ pat acc =
+    let ctx = Pat pat in
+    let {ppat_desc; ppat_loc= src} = pat in
+    match ppat_desc with
+    | Ppat_construct ({txt= Lident "[]"; loc}, None) ->
+        Cmts.relocate cmts ~src ~before:loc ~after:loc ;
+        Some (List.rev acc, loc)
+    | Ppat_construct
+        ( {txt= Lident "::"; loc}
+        , Some
+            { ppat_desc= Ppat_tuple [hd; ({ppat_attributes= []} as tl)]
+            ; ppat_loc
+            ; ppat_attributes= [] } ) ->
+        list_pat_ tl (([src; loc; ppat_loc], sub_pat ~ctx hd) :: acc)
+    | _ -> None
+  in
+  list_pat_ pat []
 
 let list_exp cmts exp =
-  let rec list_exp_ exp =
+  let rec list_exp_ exp acc =
     let ctx = Exp exp in
     let {pexp_desc; pexp_loc= src} = exp in
     match pexp_desc with
     | Pexp_construct ({txt= Lident "[]"; loc}, None) ->
         Cmts.relocate cmts ~src ~before:loc ~after:loc ;
-        Some ([], loc)
+        Some (List.rev acc, loc)
     | Pexp_construct
         ( {txt= Lident "::"; loc}
         , Some
-            {pexp_desc= Pexp_tuple [hd; tl]; pexp_loc; pexp_attributes= []}
-        ) -> (
-      match list_exp_ tl with
-      | Some (xtl, nil_loc) when List.is_empty tl.pexp_attributes ->
-          Some (([src; loc; pexp_loc], sub_exp ~ctx hd) :: xtl, nil_loc)
-      | _ -> None )
+            { pexp_desc= Pexp_tuple [hd; ({pexp_attributes= []} as tl)]
+            ; pexp_loc
+            ; pexp_attributes= [] } ) ->
+        list_exp_ tl (([src; loc; pexp_loc], sub_exp ~ctx hd) :: acc)
     | _ -> None
   in
-  let r = list_exp_ exp in
-  assert (Bool.equal (Option.is_some r) (is_sugared_list exp)) ;
```

Comment:
Why do you remove the assert?

Reply:
Because if this assert fails, the error would also be caught when checking that the ast did not changed, and this checking would provide a better error message that the assert failure.

Reply:
Great.